src/improvements/doc/NEW_TEMPLATE_GUIDE.md

```diff
@@ -0,0 +1,273 @@
+    // Parse config file
+    bytes memory configBytes = vm.parseToml(vm.readFile(taskConfigFilePath));
+    
+    // Save configuration to storage
+    for (uint256 i; i < l2Chains.length; i++) {
+        uint256 chainId = l2Chains[i].chainId;
+        taskConfig[chainId] = TaskConfig({
+            // Set template-specific parameters
+        });
+    }
+}
+```
+
+#### _build
+Implements the task logic for each chain:
+```solidity
+function _build(uint256 chainId) internal override {
+    TaskConfig memory config = taskConfig[chainId];
+    
+    // Implement task-specific logic
+    // Use config parameters to make state changes
+    // Handle any necessary contract interactions
+}
+```
+
+#### _validate
+Verifies the task executed correctly:
+```solidity
+function _validate(uint256 chainId) internal override {
+    TaskConfig memory config = taskConfig[chainId];
+        
+    // Verify state changes using assertEq, assertTrue, assertFalse foundry test functions
+    // make sure to user verbose error messages for better debugging
+}
+```
+
+## Configuration Structure
+
+### 1. TaskConfig Struct
+
+Define all parameters needed for the task:
+
+```solidity
+struct TaskConfig {
+    // Base parameters
+    uint256 chainId;
+    
+    // Template-specific parameters
+    uint256 gasLimit;
+    bytes32 configHash;
+    // Add other required fields
+}
+```
+
+### 2. TOML Configuration
+
+Create a corresponding TOML structure:
+
+```toml
+templateName = "YourTemplateName"
+
+l2chains = [
+    {name = "Chain1", chainId = 123},
+    {name = "Chain2", chainId = 456}
+]
+
+[templateConfig]
+parameter1 = "value1"
+parameter2 = 123
+```
+
+## Cross-Network Compatibility
+
+Templates must handle:
+
+1. Network-Specific Logic
+```solidity
+if (chainId == MAINNET_CHAIN_ID) {
+    // Mainnet-specific logic
+} else if (chainId == TESTNET_CHAIN_ID) {
+    // Testnet-specific logic
+}
+```
+
+2. Address Resolution
+```solidity
+// Use registry to get network-specific addresses
+address target = _getAddress(chainId, "ContractName");
+```
+
+3. Gas Configuration
+```solidity
+// Handle different gas requirements per network
+uint256 gasLimit = _getGasLimit(chainId);
+```
+
+## Error Handling
+
+1. Input Validation
+```solidity
+require(param > 0, "Invalid parameter");
+require(address(contract) != address(0), "Invalid contract");
+```
+
+2. State Validation
+```solidity
+require(
+    contract.getValue() == expectedValue,
+    "Validation failed: incorrect value"
+);
+```
+
+3. Network Checks
+```solidity
+require(
+    _isValidChainId(chainId),
+    "Unsupported network"
+);
+```
+
+## Best Practices
+
+1. Code Organization
+- Clear function names
+- Good code comments
+- Logical grouping of functionality
+
+2. Security
+- Access control checks
+- Input validation
+- State change verification
+
+3. Documentation
+- Clear parameter descriptions
+- Usage examples
+
+## Example Implementation
```

Comment:
I don't think we need an example implementation, we can point to the templates directory for those. This is another thing to keep up to date.